### PR TITLE
Refactor exceptions in gc

### PR DESCRIPTION
**Pull request by Ngoguey42**

### CODE REVIEW

File: src/irmin-pack/unix/gc.ml [hunk 1]
```diff
@@ -250,49 +245,59 @@ module Make (Args : Args) : S with module Args := Args = struct
...
       if already_copied_exn offset then No_follow
       else (
         transfer_exn offset (Int63.of_int length);
         Follow)
     in
-
-    (* 7.1. Transfer the [commit_key]. *)
-    let (_ : action) = f commit_key in
+    let node_key = Commit_value.node commit in
+    let (_ : action) = transfer_object_exn node_key in
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/gc.ml:289]: missing catch

File: src/irmin-pack/unix/gc.ml [hunk 2]
```diff
@@ -127,11 +117,16 @@ module Make (Args : Args) : S with module Args := Args = struct
...
   let transfer_parent_commit ~src ~dst key =
-    let off, len = direct_of_key key in
-    let buffer = Bytes.create len in
-    Fm.Suffix.read_exn src ~off ~len buffer;
-    Bytes.set buffer Hash.hash_size magic_parent;
-    Io.write_exn dst ~off ~len (Bytes.unsafe_to_string buffer)
+    let open Result_syntax in
+    let* off, len =
+      match Irmin_pack.Pack_key.inspect key with
+      | Indexed _ -> Error (`Commit_parent_key_is_indexed (string_of_key key))
```
Comment (tomjridge) [inline src/irmin-pack/unix/gc.ml:123]: Just to help my understanding: we check the key, which is the key of the parent of a commit; the parent of a commit is itself a commit; the check shows the parent key is indexed (which is normal? all commits are indexed?); but this is now an error? I naively expected the parent would always be indexed. What am I missing?

Comment (Ngoguey42) [reply src/irmin-pack/unix/gc.ml:123]: With brand new stores created under irmin 3.x, the only situation where Indexed commit occur is when the user retrieves one from its hash. Irmin 3.x pushes `Commit_v2` on disk, which reference other commits by offset.

Comment (Ngoguey42) [reply src/irmin-pack/unix/gc.ml:123]: Yes

**Review by icristescu — APPROVED** [review]

**Comment by Ngoguey42** [comment]
Can be merged when green

**Pull request merged**